micro_http/src/connection.rs

```diff
@@ -0,0 +1,634 @@
+                            break;
+                        }
+                    }
+                }
+                ConnectionState::WaitingForBody => {
+                    // If what we have just read is not enough to complete the request and
+                    // there are more bytes pertaining to the body of the request.
+                    if self.body_bytes_to_be_read > bytes_read as i32 - line_start_index as i32 {
+                        // Append everything that we read to our current incomplete body and update
+                        // `body_bytes_to_be_read`.
+                        self.body_vec
+                            .extend_from_slice(&self.buffer[line_start_index..bytes_read]);
+                        self.body_bytes_to_be_read -= bytes_read as i32 - line_start_index as i32;
+
+                        // Clear the buffer.
+                        for i in 0..1024 {
+                            self.buffer[i] = 0;
+                        }
+
+                        break;
+                    } else {
+                        // Append only the remaining necessary bytes to the body of the request.
+                        self.body_vec.extend_from_slice(
+                            &self.buffer[line_start_index
+                                ..(line_start_index + self.body_bytes_to_be_read as usize)],
+                        );
+                        line_start_index += self.body_bytes_to_be_read as usize;
+                        self.body_bytes_to_be_read = 0;
+                    }
+
+                    let request = self.request.as_mut().unwrap();
+                    // If there are no more bytes to be read for this request.
+                    if self.body_bytes_to_be_read == 0 {
+                        // Assign the body of the request.
+                        let placeholder: Vec<_> = self
+                            .body_vec
+                            .drain(..request.headers.content_length() as usize)
+                            .collect();
+                        request.body = Some(Body::new(placeholder));
+
+                        // If we read more bytes than we should have into the body of the request.
+                        if !self.body_vec.is_empty() {
+                            return Err(ConnectionError::StreamError);
+                        }
+
+                        self.state = ConnectionState::RequestReady;
+                    }
+                }
+                ConnectionState::RequestReady => {
+                    // This request is ready to be passed for handling.
+                    // Update the state machine to expect a new request and push this request into
+                    // the `parsed_requests` queue.
+                    self.state = ConnectionState::WaitingForRequestLine;
+                    self.body_bytes_to_be_read = 0;
+                    self.parsed_requests.push_back(self.request.take().unwrap());
+                }
```

Comment:
You need to move the rest of the _unprocessed_ buffer to the beginning of the buffer when resetting the state machine (`self.state = ConnectionState::WaitingForRequestLine`).

In the `ConnectionState::WaitingForRequestLine` state, you check if `bytes_read` is `1024` and error in case the `RequestLine` is not complete.
Without resetting/shifting the buffer here, this code could incorrectly error just because part of the next request comes in the same `try_read` but its `request_line` didn't fully fit.

Reply:
I have fixed this issue and added a test for this specific case.